src/Compilers/CSharp/Portable/Binder/WithExternAliasesBinder.cs

```diff
@@ -0,0 +1,132 @@
+        {
+            Debug.Assert(result.IsClear);
+
+            LookupSymbolInAliases(
+                ImmutableDictionary<string, AliasAndUsingDirective>.Empty,
+                this.ExternAliases,
+                originalBinder,
+                result,
+                name,
+                arity,
+                basesBeingResolved,
+                options,
+                diagnose,
+                ref useSiteInfo);
+        }
+
+        protected override void AddLookupSymbolsInfoInSingleBinder(LookupSymbolsInfo result, LookupOptions options, Binder originalBinder)
+        {
+            // If we are looking only for labels we do not need to search through the imports.
+            if ((options & LookupOptions.LabelsOnly) == 0)
+            {
+                AddLookupSymbolsInfoInAliases(
+                    ImmutableDictionary<string, AliasAndUsingDirective>.Empty,
+                    this.ExternAliases,
+                    result, options, originalBinder);
+            }
+        }
+
+        protected sealed override SourceLocalSymbol? LookupLocal(SyntaxToken nameToken)
+        {
+            return null;
+        }
+
+        protected sealed override LocalFunctionSymbol? LookupLocalFunction(SyntaxToken nameToken)
+        {
+            return null;
+        }
+
+        internal sealed override uint LocalScopeDepth => Binder.ExternalScope;
+
+        internal static WithExternAliasesBinder Create(SourceNamespaceSymbol declaringSymbol, CSharpSyntaxNode declarationSyntax, Binder next)
+        {
+            return new FromSyntax(declaringSymbol, declarationSyntax, next);
+        }
+
+        internal static WithExternAliasesBinder Create(ImmutableArray<AliasAndExternAliasDirective> externAliases, Binder next)
+        {
+            return new FromSymbols(externAliases, next);
+        }
+
+        private sealed class FromSyntax : WithExternAliasesBinder
+        {
+            private readonly SourceNamespaceSymbol _declaringSymbol;
+            private readonly CSharpSyntaxNode _declarationSyntax;
+            private ImmutableArray<AliasAndExternAliasDirective> _lazyExternAliases;
+
+            internal FromSyntax(SourceNamespaceSymbol declaringSymbol, CSharpSyntaxNode declarationSyntax, Binder next)
+                : base(next)
+            {
+                Debug.Assert(declarationSyntax.IsKind(SyntaxKind.CompilationUnit) || declarationSyntax.IsKind(SyntaxKind.NamespaceDeclaration));
+                _declaringSymbol = declaringSymbol;
+                _declarationSyntax = declarationSyntax;
+            }
+
+            internal override ImmutableArray<AliasAndExternAliasDirective> ExternAliases
+            {
+                get
+                {
+                    if (_lazyExternAliases.IsDefault)
+                    {
+                        ImmutableInterlocked.InterlockedInitialize(ref _lazyExternAliases, _declaringSymbol.GetExternAliases(_declarationSyntax));
```

Comment:
How do we evaluate where's a good place to cache? How does this compare to caching in `SourceNamespaceSymbol.GetExternAliases`?

Reply:
> How do we evaluate where's a good place to cache? How does this compare to caching in SourceNamespaceSymbol.GetExternAliases?

```SourceNamespaceSymbol``` tracks multiple caches, one per each single namespace declaration with imports. For each declaration it needs to find one first, than get the value from it. Saving the value here is cheaper than to do that every time.


---
In reply to: [588799241](https://github.com/dotnet/roslyn/pull/51676#discussion_r588799241) [](ancestors = 588799241)